config/config.go

```diff
@@ -170,3 +173,192 @@ func defaultRunPath() string {
 	}
 	return filepath.Join(homeDir, ".earthly/run")
 }
+
+// UpsertConfig adds or modifies the key to be the specified value.
+// This is saved to disk in your earthly config file.
+func UpsertConfig(config []byte, path, value string) ([]byte, error) {
+	base := &yaml.Node{}
+	yaml.Unmarshal(config, base)
+
+	if base.IsZero() {
+		// Empty file, or a simple comment results in a null document.
+		// Not handled well, so manufacture somewhat acceptable document
+		fullDoc := string(config) + "\n---"
+		yaml.Unmarshal([]byte(fullDoc), base)
+		base.Content = []*yaml.Node{{Kind: yaml.MappingNode}}
+	}
+
+	pathParts := splitPath(path)
+
+	err := validatePath(reflect.TypeOf(Config{}), pathParts)
+	if err != nil {
+		return []byte{}, errors.Wrap(err, "path is not valid")
+	}
+
+	yamlValue, err := valueToYaml(value)
+	if err != nil {
+		return []byte{}, errors.Wrap(err, "could not parse value")
+	}
+
+	setYamlValue(base, pathParts, yamlValue)
+
+	newConfig, err := yaml.Marshal(base)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return newConfig, nil
+}
+
+func splitPath(path string) []string {
+	// Allow quotes to group keys, since git repos are keys and have periods... this is why we dont just strings.Split
+	// If you screw up the quotes you will get a weird invalid path later.
+	re := regexp.MustCompile(`[^\."']+|"([^"]*)"|'([^']*)`)
+	pathParts := re.FindAllString(path, -1)
+
+	for i := 0; i < len(pathParts); i++ {
+		// If we did have a quoted string we need to prune it
+		pathParts[i] = strings.Trim(pathParts[i], `"`)
+	}
+
+	return pathParts
+}
+
+func validatePath(t reflect.Type, path []string) error {
+	if len(path) == 0 {
+		return nil
+	}
+
+	if t.Kind() == reflect.Map {
+		// Maps are only for git repos. Grab the kind on the other side of the map
+		// and advance; to validate the path on the other side of the repo name
+		return validatePath(t.Elem(), path[1:])
+	}
+
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		tag := field.Tag.Get("yaml")
+
+		if tag == path[0] {
+			return validatePath(field.Type, path[1:])
+		}
+	}
+
+	return fmt.Errorf("no path for %s", strings.Join(path, "."))
+}
+
+func valueToYaml(value string) (*yaml.Node, error) {
+	valueNode := &yaml.Node{}
+	if err := yaml.Unmarshal([]byte(value), valueNode); err != nil {
+		return nil, fmt.Errorf("%s is not a valid YAML value", value)
+	}
+
+	// Unfold all the yaml so its not mixed inline and flow styles in the final document
+	var fixStyling func(node *yaml.Node)
+	fixStyling = func(node *yaml.Node) {
+		node.Style = 0
+
+		for _, n := range node.Content {
+			fixStyling(n)
+		}
+	}
+	fixStyling(valueNode)
+
+	return valueNode.Content[0], nil
```

Comment:
The value is wrapped in a `DocumentNode`; and we dont need that node. We just want the juicy _actual_ values. If it parsed correctly, we will always have this value. I expect an error from unmarshal before this fails.